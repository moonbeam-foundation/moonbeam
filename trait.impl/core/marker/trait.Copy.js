(function() {var implementors = {
"account":[["impl <a class=\"trait\" href=\"https://doc.rust-lang.org/1.77.0/core/marker/trait.Copy.html\" title=\"trait core::marker::Copy\">Copy</a> for <a class=\"struct\" href=\"account/struct.AccountId20.html\" title=\"struct account::AccountId20\">AccountId20</a>"]],
"evm_tracing_events":[["impl <a class=\"trait\" href=\"https://doc.rust-lang.org/1.77.0/core/marker/trait.Copy.html\" title=\"trait core::marker::Copy\">Copy</a> for <a class=\"enum\" href=\"evm_tracing_events/evm/enum.CreateScheme.html\" title=\"enum evm_tracing_events::evm::CreateScheme\">CreateScheme</a>"],["impl <a class=\"trait\" href=\"https://doc.rust-lang.org/1.77.0/core/marker/trait.Copy.html\" title=\"trait core::marker::Copy\">Copy</a> for <a class=\"struct\" href=\"evm_tracing_events/gasometer/struct.Snapshot.html\" title=\"struct evm_tracing_events::gasometer::Snapshot\">Snapshot</a>"],["impl&lt;E: <a class=\"trait\" href=\"https://doc.rust-lang.org/1.77.0/core/marker/trait.Copy.html\" title=\"trait core::marker::Copy\">Copy</a>, T: <a class=\"trait\" href=\"https://doc.rust-lang.org/1.77.0/core/marker/trait.Copy.html\" title=\"trait core::marker::Copy\">Copy</a>&gt; <a class=\"trait\" href=\"https://doc.rust-lang.org/1.77.0/core/marker/trait.Copy.html\" title=\"trait core::marker::Copy\">Copy</a> for <a class=\"enum\" href=\"evm_tracing_events/runtime/enum.Capture.html\" title=\"enum evm_tracing_events::runtime::Capture\">Capture</a>&lt;E, T&gt;"],["impl <a class=\"trait\" href=\"https://doc.rust-lang.org/1.77.0/core/marker/trait.Copy.html\" title=\"trait core::marker::Copy\">Copy</a> for <a class=\"enum\" href=\"evm_tracing_events/gasometer/enum.GasometerEvent.html\" title=\"enum evm_tracing_events::gasometer::GasometerEvent\">GasometerEvent</a>"],["impl <a class=\"trait\" href=\"https://doc.rust-lang.org/1.77.0/core/marker/trait.Copy.html\" title=\"trait core::marker::Copy\">Copy</a> for <a class=\"struct\" href=\"evm_tracing_events/struct.StepEventFilter.html\" title=\"struct evm_tracing_events::StepEventFilter\">StepEventFilter</a>"]],
"moonbase_runtime":[["impl <a class=\"trait\" href=\"https://doc.rust-lang.org/1.77.0/core/marker/trait.Copy.html\" title=\"trait core::marker::Copy\">Copy</a> for <a class=\"enum\" href=\"moonbase_runtime/enum.RuntimeFreezeReason.html\" title=\"enum moonbase_runtime::RuntimeFreezeReason\">RuntimeFreezeReason</a>"],["impl <a class=\"trait\" href=\"https://doc.rust-lang.org/1.77.0/core/marker/trait.Copy.html\" title=\"trait core::marker::Copy\">Copy</a> for <a class=\"struct\" href=\"moonbase_runtime/struct.Runtime.html\" title=\"struct moonbase_runtime::Runtime\">Runtime</a>"],["impl <a class=\"trait\" href=\"https://doc.rust-lang.org/1.77.0/core/marker/trait.Copy.html\" title=\"trait core::marker::Copy\">Copy</a> for <a class=\"enum\" href=\"moonbase_runtime/enum.RuntimeHoldReason.html\" title=\"enum moonbase_runtime::RuntimeHoldReason\">RuntimeHoldReason</a>"],["impl <a class=\"trait\" href=\"https://doc.rust-lang.org/1.77.0/core/marker/trait.Copy.html\" title=\"trait core::marker::Copy\">Copy</a> for <a class=\"enum\" href=\"moonbase_runtime/enum.RuntimeSlashReason.html\" title=\"enum moonbase_runtime::RuntimeSlashReason\">RuntimeSlashReason</a>"],["impl <a class=\"trait\" href=\"https://doc.rust-lang.org/1.77.0/core/marker/trait.Copy.html\" title=\"trait core::marker::Copy\">Copy</a> for <a class=\"enum\" href=\"moonbase_runtime/enum.ProxyType.html\" title=\"enum moonbase_runtime::ProxyType\">ProxyType</a>"],["impl <a class=\"trait\" href=\"https://doc.rust-lang.org/1.77.0/core/marker/trait.Copy.html\" title=\"trait core::marker::Copy\">Copy</a> for <a class=\"enum\" href=\"moonbase_runtime/enum.RuntimeLockId.html\" title=\"enum moonbase_runtime::RuntimeLockId\">RuntimeLockId</a>"]],
"moonbeam_cli_opt":[["impl <a class=\"trait\" href=\"https://doc.rust-lang.org/1.77.0/core/marker/trait.Copy.html\" title=\"trait core::marker::Copy\">Copy</a> for <a class=\"enum\" href=\"moonbeam_cli_opt/enum.Sealing.html\" title=\"enum moonbeam_cli_opt::Sealing\">Sealing</a>"],["impl <a class=\"trait\" href=\"https://doc.rust-lang.org/1.77.0/core/marker/trait.Copy.html\" title=\"trait core::marker::Copy\">Copy</a> for <a class=\"enum\" href=\"moonbeam_cli_opt/enum.FrontierBackendType.html\" title=\"enum moonbeam_cli_opt::FrontierBackendType\">FrontierBackendType</a>"]],
"moonbeam_client_evm_tracing":[["impl <a class=\"trait\" href=\"https://doc.rust-lang.org/1.77.0/core/marker/trait.Copy.html\" title=\"trait core::marker::Copy\">Copy</a> for <a class=\"enum\" href=\"moonbeam_client_evm_tracing/types/enum.CallType.html\" title=\"enum moonbeam_client_evm_tracing::types::CallType\">CallType</a>"],["impl <a class=\"trait\" href=\"https://doc.rust-lang.org/1.77.0/core/marker/trait.Copy.html\" title=\"trait core::marker::Copy\">Copy</a> for <a class=\"enum\" href=\"moonbeam_client_evm_tracing/types/enum.CreateType.html\" title=\"enum moonbeam_client_evm_tracing::types::CreateType\">CreateType</a>"],["impl <a class=\"trait\" href=\"https://doc.rust-lang.org/1.77.0/core/marker/trait.Copy.html\" title=\"trait core::marker::Copy\">Copy</a> for <a class=\"enum\" href=\"moonbeam_client_evm_tracing/types/single/enum.TraceType.html\" title=\"enum moonbeam_client_evm_tracing::types::single::TraceType\">TraceType</a>"]],
"moonbeam_rpc_core_types":[["impl <a class=\"trait\" href=\"https://doc.rust-lang.org/1.77.0/core/marker/trait.Copy.html\" title=\"trait core::marker::Copy\">Copy</a> for <a class=\"enum\" href=\"moonbeam_rpc_core_types/enum.RequestBlockTag.html\" title=\"enum moonbeam_rpc_core_types::RequestBlockTag\">RequestBlockTag</a>"],["impl <a class=\"trait\" href=\"https://doc.rust-lang.org/1.77.0/core/marker/trait.Copy.html\" title=\"trait core::marker::Copy\">Copy</a> for <a class=\"enum\" href=\"moonbeam_rpc_core_types/enum.RequestBlockId.html\" title=\"enum moonbeam_rpc_core_types::RequestBlockId\">RequestBlockId</a>"]],
"moonbeam_rpc_primitives_debug":[["impl <a class=\"trait\" href=\"https://doc.rust-lang.org/1.77.0/core/marker/trait.Copy.html\" title=\"trait core::marker::Copy\">Copy</a> for <a class=\"enum\" href=\"moonbeam_rpc_primitives_debug/enum.TracerInput.html\" title=\"enum moonbeam_rpc_primitives_debug::TracerInput\">TracerInput</a>"]],
"moonbeam_rpc_trace":[["impl <a class=\"trait\" href=\"https://doc.rust-lang.org/1.77.0/core/marker/trait.Copy.html\" title=\"trait core::marker::Copy\">Copy</a> for <a class=\"struct\" href=\"moonbeam_rpc_trace/struct.CacheBatchId.html\" title=\"struct moonbeam_rpc_trace::CacheBatchId\">CacheBatchId</a>"]],
"moonbeam_runtime":[["impl <a class=\"trait\" href=\"https://doc.rust-lang.org/1.77.0/core/marker/trait.Copy.html\" title=\"trait core::marker::Copy\">Copy</a> for <a class=\"enum\" href=\"moonbeam_runtime/enum.RuntimeFreezeReason.html\" title=\"enum moonbeam_runtime::RuntimeFreezeReason\">RuntimeFreezeReason</a>"],["impl <a class=\"trait\" href=\"https://doc.rust-lang.org/1.77.0/core/marker/trait.Copy.html\" title=\"trait core::marker::Copy\">Copy</a> for <a class=\"enum\" href=\"moonbeam_runtime/enum.RuntimeHoldReason.html\" title=\"enum moonbeam_runtime::RuntimeHoldReason\">RuntimeHoldReason</a>"],["impl <a class=\"trait\" href=\"https://doc.rust-lang.org/1.77.0/core/marker/trait.Copy.html\" title=\"trait core::marker::Copy\">Copy</a> for <a class=\"enum\" href=\"moonbeam_runtime/enum.RuntimeSlashReason.html\" title=\"enum moonbeam_runtime::RuntimeSlashReason\">RuntimeSlashReason</a>"],["impl <a class=\"trait\" href=\"https://doc.rust-lang.org/1.77.0/core/marker/trait.Copy.html\" title=\"trait core::marker::Copy\">Copy</a> for <a class=\"enum\" href=\"moonbeam_runtime/enum.RuntimeLockId.html\" title=\"enum moonbeam_runtime::RuntimeLockId\">RuntimeLockId</a>"],["impl <a class=\"trait\" href=\"https://doc.rust-lang.org/1.77.0/core/marker/trait.Copy.html\" title=\"trait core::marker::Copy\">Copy</a> for <a class=\"enum\" href=\"moonbeam_runtime/enum.ProxyType.html\" title=\"enum moonbeam_runtime::ProxyType\">ProxyType</a>"],["impl <a class=\"trait\" href=\"https://doc.rust-lang.org/1.77.0/core/marker/trait.Copy.html\" title=\"trait core::marker::Copy\">Copy</a> for <a class=\"struct\" href=\"moonbeam_runtime/struct.Runtime.html\" title=\"struct moonbeam_runtime::Runtime\">Runtime</a>"]],
"moonbeam_runtime_common":[["impl&lt;const LOWER: <a class=\"primitive\" href=\"https://doc.rust-lang.org/1.77.0/std/primitive.u128.html\">u128</a>, const UPPER: <a class=\"primitive\" href=\"https://doc.rust-lang.org/1.77.0/std/primitive.u128.html\">u128</a>&gt; <a class=\"trait\" href=\"https://doc.rust-lang.org/1.77.0/core/marker/trait.Copy.html\" title=\"trait core::marker::Copy\">Copy</a> for <a class=\"struct\" href=\"moonbeam_runtime_common/types/struct.BoundedU128.html\" title=\"struct moonbeam_runtime_common::types::BoundedU128\">BoundedU128</a>&lt;LOWER, UPPER&gt;"]],
"moonriver_runtime":[["impl <a class=\"trait\" href=\"https://doc.rust-lang.org/1.77.0/core/marker/trait.Copy.html\" title=\"trait core::marker::Copy\">Copy</a> for <a class=\"enum\" href=\"moonriver_runtime/enum.RuntimeSlashReason.html\" title=\"enum moonriver_runtime::RuntimeSlashReason\">RuntimeSlashReason</a>"],["impl <a class=\"trait\" href=\"https://doc.rust-lang.org/1.77.0/core/marker/trait.Copy.html\" title=\"trait core::marker::Copy\">Copy</a> for <a class=\"enum\" href=\"moonriver_runtime/enum.RuntimeHoldReason.html\" title=\"enum moonriver_runtime::RuntimeHoldReason\">RuntimeHoldReason</a>"],["impl <a class=\"trait\" href=\"https://doc.rust-lang.org/1.77.0/core/marker/trait.Copy.html\" title=\"trait core::marker::Copy\">Copy</a> for <a class=\"enum\" href=\"moonriver_runtime/enum.ProxyType.html\" title=\"enum moonriver_runtime::ProxyType\">ProxyType</a>"],["impl <a class=\"trait\" href=\"https://doc.rust-lang.org/1.77.0/core/marker/trait.Copy.html\" title=\"trait core::marker::Copy\">Copy</a> for <a class=\"enum\" href=\"moonriver_runtime/enum.RuntimeLockId.html\" title=\"enum moonriver_runtime::RuntimeLockId\">RuntimeLockId</a>"],["impl <a class=\"trait\" href=\"https://doc.rust-lang.org/1.77.0/core/marker/trait.Copy.html\" title=\"trait core::marker::Copy\">Copy</a> for <a class=\"struct\" href=\"moonriver_runtime/struct.Runtime.html\" title=\"struct moonriver_runtime::Runtime\">Runtime</a>"],["impl <a class=\"trait\" href=\"https://doc.rust-lang.org/1.77.0/core/marker/trait.Copy.html\" title=\"trait core::marker::Copy\">Copy</a> for <a class=\"enum\" href=\"moonriver_runtime/enum.RuntimeFreezeReason.html\" title=\"enum moonriver_runtime::RuntimeFreezeReason\">RuntimeFreezeReason</a>"]],
"pallet_evm_precompile_batch":[["impl <a class=\"trait\" href=\"https://doc.rust-lang.org/1.77.0/core/marker/trait.Copy.html\" title=\"trait core::marker::Copy\">Copy</a> for <a class=\"enum\" href=\"pallet_evm_precompile_batch/enum.Mode.html\" title=\"enum pallet_evm_precompile_batch::Mode\">Mode</a>"]],
"pallet_parachain_staking":[["impl&lt;T: <a class=\"trait\" href=\"https://doc.rust-lang.org/1.77.0/core/marker/trait.Copy.html\" title=\"trait core::marker::Copy\">Copy</a>&gt; <a class=\"trait\" href=\"https://doc.rust-lang.org/1.77.0/core/marker/trait.Copy.html\" title=\"trait core::marker::Copy\">Copy</a> for <a class=\"struct\" href=\"pallet_parachain_staking/inflation/struct.Range.html\" title=\"struct pallet_parachain_staking::inflation::Range\">Range</a>&lt;T&gt;"],["impl <a class=\"trait\" href=\"https://doc.rust-lang.org/1.77.0/core/marker/trait.Copy.html\" title=\"trait core::marker::Copy\">Copy</a> for <a class=\"enum\" href=\"pallet_parachain_staking/types/enum.CollatorStatus.html\" title=\"enum pallet_parachain_staking::types::CollatorStatus\">CollatorStatus</a>"],["impl&lt;B: <a class=\"trait\" href=\"https://doc.rust-lang.org/1.77.0/core/marker/trait.Copy.html\" title=\"trait core::marker::Copy\">Copy</a>&gt; <a class=\"trait\" href=\"https://doc.rust-lang.org/1.77.0/core/marker/trait.Copy.html\" title=\"trait core::marker::Copy\">Copy</a> for <a class=\"enum\" href=\"pallet_parachain_staking/types/enum.DelegatorAdded.html\" title=\"enum pallet_parachain_staking::types::DelegatorAdded\">DelegatorAdded</a>&lt;B&gt;"],["impl&lt;BlockNumber: <a class=\"trait\" href=\"https://doc.rust-lang.org/1.77.0/core/marker/trait.Copy.html\" title=\"trait core::marker::Copy\">Copy</a>&gt; <a class=\"trait\" href=\"https://doc.rust-lang.org/1.77.0/core/marker/trait.Copy.html\" title=\"trait core::marker::Copy\">Copy</a> for <a class=\"struct\" href=\"pallet_parachain_staking/types/struct.RoundInfo.html\" title=\"struct pallet_parachain_staking::types::RoundInfo\">RoundInfo</a>&lt;BlockNumber&gt;"],["impl&lt;Balance: <a class=\"trait\" href=\"https://doc.rust-lang.org/1.77.0/core/marker/trait.Copy.html\" title=\"trait core::marker::Copy\">Copy</a>&gt; <a class=\"trait\" href=\"https://doc.rust-lang.org/1.77.0/core/marker/trait.Copy.html\" title=\"trait core::marker::Copy\">Copy</a> for <a class=\"struct\" href=\"pallet_parachain_staking/types/struct.CandidateBondLessRequest.html\" title=\"struct pallet_parachain_staking::types::CandidateBondLessRequest\">CandidateBondLessRequest</a>&lt;Balance&gt;"]],
"pallet_xcm_transactor":[["impl <a class=\"trait\" href=\"https://doc.rust-lang.org/1.77.0/core/marker/trait.Copy.html\" title=\"trait core::marker::Copy\">Copy</a> for <a class=\"struct\" href=\"pallet_xcm_transactor/relay_indices/struct.RelayChainIndices.html\" title=\"struct pallet_xcm_transactor::relay_indices::RelayChainIndices\">RelayChainIndices</a>"]],
"summarize_precompile_checks":[["impl <a class=\"trait\" href=\"https://doc.rust-lang.org/1.77.0/core/marker/trait.Copy.html\" title=\"trait core::marker::Copy\">Copy</a> for <a class=\"enum\" href=\"summarize_precompile_checks/enum.Network.html\" title=\"enum summarize_precompile_checks::Network\">Network</a>"],["impl <a class=\"trait\" href=\"https://doc.rust-lang.org/1.77.0/core/marker/trait.Copy.html\" title=\"trait core::marker::Copy\">Copy</a> for <a class=\"enum\" href=\"summarize_precompile_checks/enum.Format.html\" title=\"enum summarize_precompile_checks::Format\">Format</a>"]]
};if (window.register_implementors) {window.register_implementors(implementors);} else {window.pending_implementors = implementors;}})()