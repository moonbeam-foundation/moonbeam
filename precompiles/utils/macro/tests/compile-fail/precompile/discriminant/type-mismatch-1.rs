// Copyright 2019-2022 PureStake Inc.
// This file is part of Moonbeam.

// Moonbeam is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.

// Moonbeam is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.

// You should have received a copy of the GNU General Public License
// along with Moonbeam.  If not, see <http://www.gnu.org/licenses/>.

use core::marker::PhantomData;

pub struct Precompile<R>(PhantomData<R>);

#[precompile_utils_macro::precompile]
#[precompile::precompile_set]
impl<R> Precompile<R> {
	#[precompile::discriminant]
	fn discriminant(address: H160) -> Option<u64> {
		Some(42)
	}

	#[precompile::public("foo()")]
	fn foo(_discriminant: u32, test: &mut impl PrecompileHandle) -> EvmResult {
		todo!()
	}
}

fn main() { }