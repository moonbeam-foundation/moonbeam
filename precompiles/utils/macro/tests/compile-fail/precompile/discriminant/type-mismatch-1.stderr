error: All discriminants must have the same type (found u64 before)
  --> tests/compile-fail/precompile/discriminant/type-mismatch-1.rs:30:24
   |
30 |     fn foo(_discriminant: u32, test: &mut impl PrecompileHandle) -> EvmResult {
   |                           ^^^
